api/use_cases.py

```diff
@@ -41,3 +43,20 @@ def execute(self, url, predicate):
         if not subscriptions_by_url:
             raise SubscriptionNotFound()
         self.subscriptions_repo.bulk_delete([pattern.to_key(url)])
+
+
+class PublishStatusChangeUseCase:
+    """
+    Given message status changed
+    message id should be sent for notification
+    """
+
+    def __init__(self, notification_repo: repos.NotificationsRepo):
+        self.notifications_repo = notification_repo
+
+    def publish(self, message: models.Message):
+        job_payload = {
+            'predicate': f"message.{message.id}",
```

Comment:
Not sure of the use of the word 'predicate' here. This is the `key`, right?

The publisher knows that it is publishing IDs and the spreader processor expects the key to be IDs, so I don't think we need to be specific in the dictionaries. Unless you think it should work differently to that.

(Side note, the code in `libtrustbridge/websub/repos.py:SubscriptionsRepo` shouldn't refer to predicates, only patterns. Very similar, but distinct things).

Reply:
may be it's `topic`, not the key; but generally I agree.

even if the value is "predicate" for some websub users it still "topic" from the websub perspective

Reply:
I'm a bit confused about what `predicate` really means, previously I defined a [Pattern](https://github.com/trustbridge/libtrustbridge/blob/master/libtrustbridge/websub/domain.py#L24
)` model that takes predicate and produces a key. Then a Node can subscribe by pattern

Reply:
Predicate is a business term which is something like `UN.Cefact.Certificate.created`. We need to subscribe to 2 kinds of things in general: that predicates (so somebody receives news about certificates created by UN.Cefact.Certificate.created subscription, about anything using `UN.Cefact.*`) and to mesages (so I subscribe to `message.*.created` or `message.{my_id}.status`). So for this PR (I believe) it's enough not to think too much about the predicate meaning, leaving it to users and receives, and just consider it to be a dot-separated value.

Reply:
Yes, topic, not key. My bad.

Predicate is a business term, but this GENERIC subscription code (everything in libtrustbridge) shouldn't know anything about predicates or messages or statuses. It supports two styles of subscription: exact match (eg. ID) and pattern wildcard (A.B.C.*).

The use cases we have are "I want to know everything about the object with this ID" and "I want to know everything about any object that matches this pattern A.B.C" (where "pattern" means that A.B.C matches A.B.C.D but not A.B.CD.D - ie. it's slightly more complicated than simple wildcard matching). This is all the generic code in libtrustbridge needs to handle.

> previously I defined a Pattern` model that takes predicate and produces a key

I guess, in short, I am trying to say that the Pattern model shouldn't take a predicate but a dot separated pattern. It's a technicality really, but one that keeps things simpler in my mind.

The hub endpoint, validation of whether a topic is valid and what to publish is all contained within the node or channel code, each of which DO understand their domain objects - messages, predicates, etc...

Now, re: `message.{id}.status` - I'm not sure about the need for `.status` on the end, but I don't see how it hurts. Also not sure about the need for `message.` at the start. As long as the subscriber (node) and publisher (channel) use the same convention, then it doesn't matter. Our node won't want to subscribe to `.created` only events - it will always want to know about ANY change to the message that the channel wants to publish. At the moment, any change is almost certainly limited to status changes only.

🐥

Reply:
So we rename "predicate" to "Pattern" and just use it as predicate on business code and as pattern on low-level subscription one?

My idea about the `message.{id}.*` was just to separate it from UN predicates. How else will you subscribe to the specific message ID updates? ("Im a chambers app and I want to get news about my message which I've just sent - is it accepted or rejected")

Reply:
Re: predicate/pattern - yep, exactly.

Subscribing to a message ID is a separate endpoint, hub, flask view, with it's own logic. My idea was that either we use two separate s3 services or a different prefix path in the same s3 service for different types of subscriptions. There will be different publishers for the different subscription types, but perhaps this means we also have different processors to send callbacks and perhaps that's not ideal.

There was code that illustrated the bucket prefix point, but I can't find it right now, will try later tonight.

Reply:
> "Im a chambers app and I want to get news about my message which I've just sent - is it accepted or rejected"

Just for clarity: When you get the response from publishing your message to the node, there is a header telling you the hub url that you can subscribe at.

This hub url would be different to the one where you want to find out about all messages destined TO me, not FROM me. To get THAT hub url, you might do a HEAD on `/messages/incoming/` for example (I'm sure there is a url for it already) and in the response headers would be the hub url.

Reply:
Okay, if we do a separate endpoint to subscribe to message updates and read it from the message POST response headers then I'm fine. Just please document it somewhere so I implement it in the chambers/trade apps. Now they assume that the subscription endpoint is single and just use different predicates (and node processing steps just send notifications using the same predicates later).